lemminx-liberty/src/main/java/io/openliberty/tools/langserver/lemminx/LibertyDiagnosticParticipant.java

```diff
@@ -144,15 +164,53 @@ private void validateIncludeLocation(DOMDocument domDocument, List<Diagnostic> l
                 DOMAttr optNode = node.getAttributeNode("optional");
                 if (optNode == null) {
-                    list.add(new Diagnostic(range, IMPLICIT_NOT_OPTIONAL_MESSAGE, DiagnosticSeverity.Error, "liberty-lemminx", IMPLICIT_NOT_OPTIONAL_CODE));
+                    diagnosticsList.add(new Diagnostic(range, IMPLICIT_NOT_OPTIONAL_MESSAGE, DiagnosticSeverity.Error, LIBERTY_LEMMINX_SOURCE, IMPLICIT_NOT_OPTIONAL_CODE));
                 } else if (optNode.getValue().equals("false")) {
                     Range optRange = XMLPositionUtility.createRange(optNode.getStart(), optNode.getEnd(), domDocument);
-                    list.add(new Diagnostic(optRange, NOT_OPTIONAL_MESSAGE, DiagnosticSeverity.Error, "liberty-lemminx", NOT_OPTIONAL_CODE));
+                    diagnosticsList.add(new Diagnostic(optRange, NOT_OPTIONAL_MESSAGE, DiagnosticSeverity.Error, LIBERTY_LEMMINX_SOURCE, NOT_OPTIONAL_CODE));
                 }
-                list.add(new Diagnostic(range, MISSING_FILE_MESSAGE, DiagnosticSeverity.Warning, "liberty-lemminx", MISSING_FILE_CODE));
+                diagnosticsList.add(new Diagnostic(range, MISSING_FILE_MESSAGE, DiagnosticSeverity.Warning, LIBERTY_LEMMINX_SOURCE, MISSING_FILE_CODE));
             }
         } catch (IllegalArgumentException e) {
-            list.add(new Diagnostic(range, MISSING_FILE_MESSAGE, DiagnosticSeverity.Warning, "liberty-lemminx-exception", MISSING_FILE_CODE));
+            diagnosticsList.add(new Diagnostic(range, MISSING_FILE_MESSAGE, DiagnosticSeverity.Warning, "liberty-lemminx-exception", MISSING_FILE_CODE));
+        }
+    }
+
+    /**
+     * Create temporary diagnostics for validation for single pass-through.
+     * @param domDocument
+     * @param diagnosticsList
+     * @param configElementNode
+     * @param tempDiagnosticsList
+     */
+    private void holdConfigElement(DOMDocument domDocument, DOMNode configElementNode, List<Diagnostic> tempDiagnosticsList) {
+        String configElementName = configElementNode.getNodeName();
+        Range range = XMLPositionUtility
+                .createRange(configElementNode.getStart(), configElementNode.getEnd(), domDocument);
+        Diagnostic tempDiagnostic = new Diagnostic(range, MISSING_CONFIGURED_FEATURE_MESSAGE, null, LIBERTY_LEMMINX_SOURCE, MISSING_CONFIGURED_FEATURE_CODE);
+        tempDiagnostic.setSource(configElementName);
+        tempDiagnosticsList.add(tempDiagnostic);
+    }
+
+    /**
+     * Compare the required feature set with included feature set for each config element.
+     * @param diagnosticsList
+     * @param tempDiagnosticsList
+     * @param featureGraph
+     */
+    private void validateConfigElements(List<Diagnostic> diagnosticsList, List<Diagnostic> tempDiagnosticsList, FeatureListGraph featureGraph) {
+        for (Diagnostic tempDiagnostic : tempDiagnosticsList) {
+            if (includedFeatures.isEmpty()) {
+                diagnosticsList.add(tempDiagnostic);
+                continue;
```

Comment:
We can actually optimize this. If `includedFeatures.isEmpty()`, it will be for every iteration. Just check before entering the `for` loop and `addAll` of the temp diagnostics to the list and return.